### PR TITLE
Bugfix/sdb 9074 si toolkit benchmarking reports success even when something fails

### PR DESCRIPTION
**Pull request by saidkarim**
After the PR is approved, I'll format **s64da_benchmark_toolkit/prepare.py** with black

### CODE REVIEW

**Review by ol-zinch — APPROVED** [review]

File: s64da_benchmark_toolkit/prepare.py [hunk 1]
```diff
@@ -105,21 +116,23 @@ def _run_shell_task(self, task, return_output=False):
 
             if return_output:
                 stdout, _ = p.communicate()
-                return stdout
+                print(f"output of run_shell_task: {stdout.decode('utf-8').strip()}")
```
Comment (oferz) [inline s64da_benchmark_toolkit/prepare.py:119]: I think you should simply print out the output without any prefixes. `run_shell_task` doesn't mean anything to the end users, if anything they only expect to see the output from the command.
I would also remove the `strip()` both from here and from the return command below.

File: s64da_benchmark_toolkit/prepare.py [hunk 2]
```diff
@@ -105,21 +116,23 @@ def _run_shell_task(self, task, return_output=False):
...
 
         # randomize the tasks to decrease lock contention
         random.shuffle(tasks)
         with ThreadPoolExecutor(max_workers=self.args.max_jobs) as executor:
             futures = [get_future(executor, task) for task in tasks]
             for completed_future in as_completed(futures):
                 exc = completed_future.exception()
+                PrepareBenchmarkFactory.check_ingest(completed_future.result())
                 if exc:
                     print(f'Task threw an exception: {exc}')
```
Comment (oferz) [inline s64da_benchmark_toolkit/prepare.py:137]: `check_ingest` will raise an error if the ingest failed, so this should be gracefully handled. Meaning you should put it in `try/except NoIngestException` clause. Something like:
```
                exc = completed_future.exception()
                ingest_check_failed = False
                try:
                    PrepareBenchmarkFactory.check_ingest(completed_future.result())
                except NoIngestException:
                    ingest_check_failed = True
                if exc:
                    print(f'Task threw an exception: {exc}')
                    print_tb(exc.__traceback__)
                if exc or ingest_check_failed:
                    for future in futures:
                        future.cancel()
                    exit(1)
```

**Review by oferz — CHANGES_REQUESTED** [review]
Can you also create a unit test for this?

File: s64da_benchmark_toolkit/prepare.py [hunk 3]
```diff
@@ -21,6 +21,9 @@
 
 s64_benchmark_toolkit_root_dir = Path(os.path.abspath(__file__)).parents[1]
 
+class NoIngestException(Exception):
+    """Exception when no rows inserted into the table."""
```
Comment (0xfeedface) [inline s64da_benchmark_toolkit/prepare.py:25]: s/inserted/have been inserted

File: s64da_benchmark_toolkit/prepare.py (hunk 1 above)
Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: done

File: s64da_benchmark_toolkit/prepare.py (hunk 2 above)
Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: Done

File: s64da_benchmark_toolkit/prepare.py (hunk 3 above)
Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: Done

File: s64da_benchmark_toolkit/prepare.py [hunk 4]
```diff
@@ -74,191 +82,228 @@ def supports_cluster(self):
...
             p.wait()
-            if(p.returncode != 0):
+            if p.returncode != 0:
                 self.cancel_event.set()
                 exit(task)
 
             if return_output:
                 stdout, _ = p.communicate()
-                return stdout
+                print(f"{stdout.decode('utf-8')}")
```
Comment (oferz) [inline s64da_benchmark_toolkit/prepare.py:132]: Why do we need the `f""` here?

File: s64da_benchmark_toolkit/prepare.py [hunk 5]
```diff
@@ -44,24 +49,27 @@ class PrepareBenchmarkFactory:
     CLUSTER_SPEC = {}
 
     def __init__(self, args, benchmark):
-        schemas_dir = os.path.join(s64_benchmark_toolkit_root_dir, benchmark.base_dir, "schemas")
+        schemas_dir = os.path.join(
```
Comment (oferz) [inline s64da_benchmark_toolkit/prepare.py:52]: Can you please undo the formatting changes? They just make the PR look bigger than it is, and they are probably not consistent with the styling of the rest of the code.
We should probably tackle styling of the code, but it should be done independently.

Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: Done. I hoped to do formatting gradually.

File: s64da_benchmark_toolkit/prepare.py (hunk 4 above)
Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: Done

**Comment by saidkarim** [comment]
> Can you also create a unit test for this?

It's very difficult. Because the data generation and copy command come together from prepare.py. I don't know how to separate them. Returning zero from this command in a unit test for me doesn't make sense.

File: s64da_benchmark_toolkit/prepare.py [hunk 6]
```diff
@@ -105,27 +116,34 @@ def _run_shell_task(self, task, return_output=False):
 
             if return_output:
                 stdout, _ = p.communicate()
-                return stdout
+                print({stdout.decode('utf-8')})
```
Comment (oferz) [inline s64da_benchmark_toolkit/prepare.py:119]: You forgot to remove the `{}`

Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: Done. )) I also started TC job

Comment (saidkarim) [reply s64da_benchmark_toolkit/prepare.py]: TC successfully finished ))

**Review by oferz — APPROVED** [review]

**Pull request merged**